crates/ruff/src/rules/flake8_todo/rules.rs

```diff
@@ -0,0 +1,384 @@
+            prev_token_is_todo = false;
+            continue;
+        };
+
+        // Check for errors on the tag: TDO001/TDO006.
+        // Unwrap is safe because the "tag" capture group is required to get here.
+        let captures = captures_opt.peek().unwrap();
+        let tag = captures.name("tag").unwrap().as_str();
+        if tag != "TODO" {
+            diagnostics_ref.push(Diagnostic::new(
+                InvalidTodoTag {
+                    tag: String::from(tag),
+                },
+                range,
+            ));
+
+            if tag.to_uppercase() == "TODO" {
+                let invalid_capitalization = Diagnostic::new(
+                    InvalidCapitalizationInTodo {
+                        tag: String::from(tag),
+                    },
+                    range,
+                )
+                .with_fix(
+                    if should_autofix(autofix, settings, Rule::InvalidCapitalizationInTodo) {
+                        let first_t_position = find_first_t_position(comment);
+
+                        Fix::new(vec![Edit::replacement(
+                            "TODO".to_string(),
+                            Location::new(range.location.row(), first_t_position),
+                            Location::new(range.location.row(), first_t_position + TODO_LENGTH),
+                        )])
+                    } else {
+                        Fix::empty()
+                    },
+                );
+
+                diagnostics_ref.push(invalid_capitalization);
+            }
+        }
+
+        // Check the rest of the capture groups for errors
+        for capture_group_index in 2..=NUM_CAPTURE_GROUPS {
+            if captures.get(capture_group_index).is_some() {
+                continue;
+            }
+
+            if let Some(diagnostic) = get_regex_error(capture_group_index, &range, diagnostics_ref)
+            {
+                diagnostics_ref.push(diagnostic);
+            };
+        }
+
+        prev_token_is_todo = true;
+        prev_token_todo_start = find_first_t_position(comment);
+    }
+
+    diagnostics
+}
+
+/// Mapper for static regex errors caused by a capture group at index i (i > 1 since the tag
+/// capture group could lead to multiple diagnostics being pushed)
+fn get_regex_error(i: usize, range: &Range, diagnostics: &mut [Diagnostic]) -> Option<Diagnostic> {
+    match i {
+        2usize => Some(Diagnostic::new(MissingAuthorInTodo, *range)),
+        3usize => Some(Diagnostic::new(MissingColonInTodo, *range)),
+        4usize => {
+            if diagnostics
+                .last()
+                .map_or(true, |last| last.kind != MissingColonInTodo.into())
+            {
+                Some(Diagnostic::new(MissingSpaceAfterColonInTodo, *range))
+            } else {
+                None
+            }
+        }
+        5usize => Some(Diagnostic::new(MissingTextInTodo, *range)),
+        _ => None,
+    }
```

Comment:
I wonder if it would improve performance if we avoid capture groups altogether by:

* Using a Regex to find any occurrences of `TODO`, `FIXME` or `BUG` (`regex.find(text)`)
* Then use `str` methods to test for author, colon, space, etc). 

I might be wrong but my impression is that it shouldn't be too hard to write and it has the benefit that we have fewer regex (that I personally find hard to understand and maintain) and less capture group matching. What do you think?

Reply:
I can definitely look into it. My worry with `str` methods is that the rules may not all be applied together, so testing for stuff may get tricky. I'm sure there's a way, but I'll have to check it out. 

For example, testing for a colon in `# TODO: test`, `# TODO we need to fix a few things: a, b, and c` and `# TODO (evanrittenhouse) test` is tough because you don't know where the colon should be. Should it be after the closing parenthesis of an author tag - that may not exist. We can't map it off of an offset/column since we don't know how long the author/tag is. We also can't just test if a colon is in the string since it could be in the TODO's text. 

Again, I'm sure there's a way, and I agree that it'd be easier to reason about than regex capture groups. Just haven't had a chance to sit down and think about what the detection code would look like :)

Reply:
> For example, testing for a colon in # TODO: test, # TODO we need to fix a few things: a, b, and c and # TODO (evanrittenhouse) test 

My understanding is that the colon should either come directly after the `TODO` or after the author. Is this correct? If so, then we would need to skip the "author" part even if the author rule is disabled. This isn't any different than what the `Regex` is doing today, it just happens implicitly. 

I'm also OK merging as is and playing around with string methods as a separate PR.

Reply:
Yeah, it'd just be copying the `Regex` into code. If it's cool with you, I'll probably make the speed adjustments/`Peekable` changes as mentioned above, then try to merge?

I've noted down the string method stuff in my to-do list, so I'll check that out when I get some time!

Reply:
Sounds good to me. Thank you